### PR TITLE
Guarantee ordering for SignalR client HttpConnection.send #10780

### LINKED ISSUE #9901: Blazor (server-side) System.InvalidOperationException: Received a notification for a rendered batch when not expecting it.
I`ve got an Blazor server-side app deployed as self-contained (dotnet 3.0 preview4) on Azure AppServices.
It seems to run fine on simple Pages (counter with additional buttons or the weather forecast displayed in an pageable datagrid), but it dies after 0-5 UI interactions on a more advanced page. 
In the browser-console the following error occurs and after that error the signalR connection is disconnected (without the automatic reconnection kicking in):
grafik

On my local dev maschine no such error occured even once.

To see this error happening open this [link to the deployed webApp](http://miniaturegolf.azurewebsites.net/MiniatureGolf) and click on some button or dropdowns, after some UI interactions the error should occur in the browser console.

If needed the source code for this app is located [here](https://github.com/marcelgoldstein/MiniatureGolf).

Same error as reported in this comments:
[https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-core-3-0-preview-4/#comments-367](https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-core-3-0-preview-4/#comments-367)

### PR DESCRIPTION
**Pull request by pranavkm**
Resolves #9901

### CODE REVIEW

File: eng/targets/Npm.Common.targets [hunk 1]
```diff
@@ -47,13 +48,13 @@
...
   <Target Name="Pack" Condition="'$(IsPackable)' == 'true'" DependsOnTargets="$(PackDependsOn)">
     <PropertyGroup>
-      <_BackupPackageJson>$(IntermediateOutputPath)$(MSBuildProjectName).package.json.bak</_BackupPackageJson>
       <_PackageTargetPath>$(MSBuildProjectDirectory)\$(PackageFileName)</_PackageTargetPath>
     </PropertyGroup>
 
     <Copy SourceFiles="$(PackageJson)" DestinationFiles="$(_BackupPackageJson)" />
 
     <Yarn Command="version --no-git-tag-version --new-version $(PackageVersion)" />
+    <Exec Command="node $(MSBuildThisFileDirectory)..\scripts\update-packagejson-links.js $(PackageJson) $(PackageVersion)" />
```
Comment (pranavkm) [inline eng/targets/Npm.Common.targets]: @anurse

File: src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj [hunk 2]
```diff
@@ -7,23 +7,10 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <WebpackInputs Include="src\**\*.ts" />
```
Comment (pranavkm) [inline src/Components/Browser.JS/Microsoft.AspNetCore.Components.Browser.JS.npmproj]: I considered adding build incrementalism to the root Npm.Common.targets and got something working fairly well. Might be something to consider to make inner loops with `dotnet build` be nice.

File: src/SignalR/clients/ts/signalr/src/HttpConnection.ts [hunk 3]
```diff
@@ -501,3 +515,103 @@ export class HttpConnection implements IConnection {
...
+        if (!this.transportResult) {
+            this.transportResult = new PromiseSource();
+        }
+        return this.transportResult.promise;
+    }
+
+    public stop(): Promise<void> {
+        this.executing = false;
+        this.sendBufferedData.resolve();
+        return this.sendLoopPromise;
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HttpConnection.ts]: Thanks for remembering.

File: src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts [hunk 4]
```diff
@@ -844,3 +845,232 @@ describe("HttpConnection", () => {
...
+                await promiseSource;
+            });
+        transport.send = sendMock;
+
+        const queue = new TransportSendQueue(transport);
+
+        const send = queue.send("Test");
+        await queue.stop();
+
+        await expect(send).rejects.toBe("Connection stopped.");
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/tests/HttpConnection.test.ts]: Awesomesauce

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
I only evaluated the runtime changes, not the build changes.

File: src/Components/Browser.JS/package.json [hunk 5]
```diff
@@ -5,14 +5,15 @@
   "description": "",
   "main": "index.js",
   "scripts": {
+    "prebuild": "yarn install --mutex network",
```
Comment (natemcmaster) [inline src/Components/Browser.JS/package.json]: Curious - why was this added? While I think it's normal for someone who is used to `dotnet build` to expect `yarn run build` to also restore, I don't know if this is something that's normally done in NodeJS projects. Yarn is pretty good about no-op and caching, but let's be careful about unintended consequences for running restore during build.

File: eng/scripts/update-packagejson-links.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+const fs = require("fs");
+
+const args = process.argv.slice(2);
+const packageJson = args[0];
+const packageVersion = args[1];
+
+const fileContent = fs.readFileSync(packageJson);
+const updatedContent = fileContent.toString().replace(/\"link:.*\"/g, `">=${packageVersion}"`);
+if (fileContent != updatedContent) {
+    fs.writeFileSync(packageJson, updatedContent);
```
Comment (natemcmaster) [inline eng/scripts/update-packagejson-links.js]: 👏  much better than trying to do this with MSBuild.

File: build/repo.props [hunk 7]
```diff
@@ -126,7 +126,6 @@
         <NodeJsProjects Include="
                           $(RepoRoot)src\SignalR\**\*.npmproj;
                           $(RepoRoot)src\Middleware\**\*.npmproj;
-                          $(RepoRoot)src\Components\Browser.JS\**\*.npmproj;
```
Comment (natemcmaster) [inline build/repo.props]: Why is this being removed from the build?

Comment (pranavkm) [reply build/repo.props]: I'm not a 100% certain since I didn't investigate a great deal as to why, but from what I could guess building the SignalR functional test project and the Browser.JS project in parallel would party over the same set of linked directories. However, we don't actually need to build the Browser.JS project on the CI - we have a CodeCheck to verify it is up to date and the generated files are checked in.

File: eng/scripts/update-packagejson-links.js (hunk 6 above)
Comment (pranavkm) [reply eng/scripts/update-packagejson-links.js]: My only concern was that we now invoke `node` in MSBuild. But presumably we already do that as part of `yarn`

File: src/Components/Browser.JS/package.json (hunk 5 above)
Comment (pranavkm) [reply src/Components/Browser.JS/package.json]: Think I could remove it now. When I was messing around with getting the linked build, I experimented with differing the `yarn install` until after project discovery. I'll try and undo.

File: src/SignalR/clients/ts/signalr/package.json [hunk 8]
```diff
@@ -46,6 +46,7 @@
   ],
   "devDependencies": {
     "@types/eventsource": "^1.0.2",
+    "@types/jest": "^23.3.2",
```
Comment (pranavkm) [inline src/SignalR/clients/ts/signalr/package.json]: Btw, this gives me typing and completion for `jest` APIs in the SignalR test projects.
